### PR TITLE
smart filter created

### PR DESCRIPTION
**Pull request by vp9920**
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23discussion_r49666132%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23discussion_r49666883%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171473789%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171482166%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171482429%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171482840%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171673934%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171723873%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171754286%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171767831%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171810481%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23discussion_r49799902%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23discussion_r49800187%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171954775%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23issuecomment-171473789%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22filter%20is%20a%20really%20great%20feature.%20have%20no%20comments%20and%20corrections.%20your%20code%2C%20your%20unit%20tests.%20entropy%20has%20not%20increased.%20And%20that%27s%20good.%5Cr%5Cnexcept%20few%20annoying%20%28from%20the%20user%27s%20point%29%20things.%5Cr%5Cn%5Cr%5Cn%5Cr%5Cnconsider%20the%20following%20scenario%3A%5Cr%5Cn1.%20Enter%20%60au%60...%20you%27ll%20see%20the%20Author%20typeahead%5Cr%5Cn2.%20Choose%20author%20from%20typeahead%20...%20you%20will%20not%20see%20the%20hints%3A%20%60auithor%3A%60%20and%20%60author%3D%60%20%28you%20will%20see%20hints%20if%20you%20enter%20text%20from%20the%20keyboard%20or%20after%20the%20next%20action%29%5Cr%5Cnso%20bad%20things%20are%3A%20no%20hints%2C%20and%20not%20too%20intuitive%20difference%20between%20%60%3A%60%20and%20%60%3D%60%20%28we%20shall%20think%20about%20the%20small%20%21%5Bquestion%20sign%5D%28http%3A//forums.androidcentral.com/images/icons/icon5.png%29%20near%20the%20filter%29%5Cr%5Cn%5Cr%5CnAnd%20one%20more%3A%5Cr%5Cn-%20environment%5Cr%5Cn-%20env%5Cr%5Cn-%20envFull%5Cr%5Cn%5Cr%5Cntooo%20many%20options%21%20we%20shall%20respect%20our%20user%20%3A%29%22%2C%20%22created_at%22%3A%20%222016-01-13T23%3A45%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7273003%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok%2C%20we%20can%20exclude%20environment%2C%20but%20env%20and%20envFull%20is%20not%20identical.%5Cr%5CnAbout%20closing%20hints%20when%20user%20accept%20one%20of%20hints%20...%20I%20know%20about%20that%20and%20I%20don%27t%20like%20it%20too%2C%20but%20I%20thought%20that%20this%20behavior%20impossible%20to%20change%20..%20and%20yet%2C%20maybe%20possible%20..%20I%27ll%20try%20to%20do%20something.%22%2C%20%22created_at%22%3A%20%222016-01-14T00%3A21%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11821757%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/vp9920%22%7D%7D%2C%20%7B%22body%22%3A%20%22About%20deleting%20trello%20and%20qjuery%2C%20I%20never%20deleted%20them%2C%20this%20is%20result%20of%20merge%22%2C%20%22created_at%22%3A%20%222016-01-14T00%3A23%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11821757%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/vp9920%22%7D%7D%2C%20%7B%22body%22%3A%20%22Not%20closed%20%3Csection%3E%20-%20from%20master%22%2C%20%22created_at%22%3A%20%222016-01-14T00%3A26%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11821757%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/vp9920%22%7D%7D%2C%20%7B%22body%22%3A%20%22About%20closing%20hints%2C%20I%20tried%20all%20I%20could%2C%20I%20don%27t%20know%20how%20to%20change%20this%20behavior%22%2C%20%22created_at%22%3A%20%222016-01-14T15%3A29%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11821757%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/vp9920%22%7D%7D%2C%20%7B%22body%22%3A%20%22And%20directive%20template%20was%20replaced%20into%20separate%20file%22%2C%20%22created_at%22%3A%20%222016-01-14T18%3A02%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11821757%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/vp9920%22%7D%7D%2C%20%7B%22body%22%3A%20%22Wait%20a%20while%2C%20I%20wanna%20replace%20two-way%20data%20binding%20with%20one-way%22%2C%20%22created_at%22%3A%20%222016-01-14T19%3A35%3A27Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11821757%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/vp9920%22%7D%7D%2C%20%7B%22body%22%3A%20%22done%22%2C%20%22created_at%22%3A%20%222016-01-14T20%3A17%3A46Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11821757%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/vp9920%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3E%20About%20deleting%20trello%20and%20qjuery%2C%20I%20never%20deleted%20them%2C%20this%20is%20result%20of%20commit%20after%20resetting%20to%20updated%20master%5Cr%5Cn%5Cr%5Cnbut%20nevertheless%20your%20commit%20deletes%20that%202%20lines.%5Cr%5Cnand%20your%20branch%20has%20conflicts%20%28I%20can%27t%20merge%20it%29%20-%20We%20can%20check%20tomorrow%20if%20you%20wish%22%2C%20%22created_at%22%3A%20%222016-01-14T23%3A07%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7273003%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF%22%7D%7D%2C%20%7B%22body%22%3A%20%22filter%20constants%20at%20last%20deserved%20separate%20file%22%2C%20%22created_at%22%3A%20%222016-01-15T13%3A05%3A31Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/11821757%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/vp9920%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20b76055bf173c29c06ebee807d1e866d7085080ee%20TCMSApp/src/client/index.html%206%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23discussion_r49666883%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22deleted%20by%20mistake%3F%22%2C%20%22created_at%22%3A%20%222016-01-13T23%3A21%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7273003%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20TCMSApp/src/client/index.html%3AL39-45%22%7D%2C%20%22Pull%20b76055bf173c29c06ebee807d1e866d7085080ee%20TCMSApp/src/client/app/runs/runs.html%2086%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23discussion_r49666132%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%3C/section%3E%3E%20%3F%20%29%22%2C%20%22created_at%22%3A%20%222016-01-13T23%3A13%3A40Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7273003%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20TCMSApp/src/client/app/runs/runs.html%3AL176-180%22%7D%2C%20%22Pull%20a586e23700b80bc6ddb4f2f3322a5e973093bee4%20TCMSApp/src/client/app/core/constants.js%207%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23discussion_r49799902%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22if%20we%20will%20have%20some%20API%20%20constants%20here%20too%20this%20file%20will%20be%20too%20crowded.%20your%20filter%20settings%20deserve%20the%20separate%20file.%22%2C%20%22created_at%22%3A%20%222016-01-14T23%3A08%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7273003%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20TCMSApp/src/client/app/core/constants.js%3AL6-46%22%7D%2C%20%22Pull%20a586e23700b80bc6ddb4f2f3322a5e973093bee4%20TCMSApp/karma.conf.js%209%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-012/pull/119%23discussion_r49800187%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20see%20that%20now%20you%20have%20no%20literal%20with%20the%20template%20to%20inject%20into%20the%20test.%20%5Cr%5CnNice%21%5Cr%5CnCan%20you%20please%20describe%20the%20solution%20in%20few%20details%3F%20%3A%29%22%2C%20%22created_at%22%3A%20%222016-01-14T23%3A11%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7273003%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20TCMSApp/karma.conf.js%3AL54-64%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull b76055bf173c29c06ebee807d1e866d7085080ee TCMSApp/src/client/app/runs/runs.html 86'></a> &nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-012/pull/119#discussion_r49666132'>File: TCMSApp/src/client/app/runs/runs.html:L176-180</a></b>
- <a href='https://github.com/ITsvetkoFF'></a> </section>> ? )
- [ ] <a href='#crh-comment-Pull b76055bf173c29c06ebee807d1e866d7085080ee TCMSApp/src/client/index.html 6'></a> &nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-012/pull/119#discussion_r49666883'>File: TCMSApp/src/client/index.html:L39-45</a></b>
- <a href='https://github.com/ITsvetkoFF'></a> deleted by mistake?
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-012/pull/119#issuecomment-171473789'>General Comment</a></b>
- <a href='https://github.com/ITsvetkoFF'></a> filter is a really great feature. have no comments and corrections. your code, your unit tests. entropy has not increased. And that's good.
except few annoying (from the user's point) things.
consider the following scenario:
1. Enter `au`... you'll see the Author typeahead
2. Choose author from typeahead ... you will not see the hints: `auithor:` and `author=` (you will see hints if you enter text from the keyboard or after the next action)
so bad things are: no hints, and not too intuitive difference between `:` and `=` (we shall think about the small question sign near the filter)
And one more:
- environment
- env
- envFull
tooo many options! we shall respect our user :)
- <a href='https://github.com/vp9920'></a> Ok, we can exclude environment, but env and envFull is not identical.
About closing hints when user accept one of hints ... I know about that and I don't like it too, but I thought that this behavior impossible to change .. and yet, maybe possible .. I'll try to do something.
- <a href='https://github.com/vp9920'></a> About deleting trello and qjuery, I never deleted them, this is result of merge
- <a href='https://github.com/vp9920'></a> Not closed <section> - from master
- <a href='https://github.com/vp9920'></a> About closing hints, I tried all I could, I don't know how to change this behavior
- <a href='https://github.com/vp9920'></a> And directive template was replaced into separate file
- <a href='https://github.com/vp9920'></a> Wait a while, I wanna replace two-way data binding with one-way
- <a href='https://github.com/vp9920'></a> done
- <a href='https://github.com/ITsvetkoFF'></a> > About deleting trello and qjuery, I never deleted them, this is result of commit after resetting to updated master
but nevertheless your commit deletes that 2 lines.
and your branch has conflicts (I can't merge it) - We can check tomorrow if you wish
- <a href='https://github.com/vp9920'></a> filter constants at last deserved separate file
- [ ] <a href='#crh-comment-Pull a586e23700b80bc6ddb4f2f3322a5e973093bee4 TCMSApp/src/client/app/core/constants.js 7'></a> &nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-012/pull/119#discussion_r49799902'>File: TCMSApp/src/client/app/core/constants.js:L6-46</a></b>
- <a href='https://github.com/ITsvetkoFF'></a> if we will have some API  constants here too this file will be too crowded. your filter settings deserve the separate file.
- [ ] <a href='#crh-comment-Pull a586e23700b80bc6ddb4f2f3322a5e973093bee4 TCMSApp/karma.conf.js 9'></a> &nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-012/pull/119#discussion_r49800187'>File: TCMSApp/karma.conf.js:L54-64</a></b>
- <a href='https://github.com/ITsvetkoFF'></a> I see that now you have no literal with the template to inject into the test.
Nice!
Can you please describe the solution in few details? :)

<a href='https://www.codereviewhub.com/ITsvetkoFF/Kv-012/pull/119?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/ITsvetkoFF/Kv-012/pull/119?approve=1'></a>&nbsp;<a href='https://github.com/ITsvetkoFF/Kv-012/pull/119'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by ITsvetkoFF** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/ITsvetkoFF/Kv-012/pull/119'>Refresh</a>.<a href='#crh-update-none'></a>

File: TCMSApp/src/client/app/runs/runs.html [hunk 1]
```diff
@@ -147,4 +176,4 @@
             </div>
         </div>
     </div>
-</div>
+</section>>
```
Comment (ITsvetkoFF) [inline TCMSApp/src/client/app/runs/runs.html]: </section>> ? )

File: TCMSApp/src/client/index.html [hunk 2]
```diff
@@ -39,10 +39,6 @@
         </div>
     </div>
 
-
-    <script src="//code.jquery.com/jquery-1.11.3.min.js"></script>
-    <script src="https://api.trello.com/1/client.js?key=50c269d155e10e2d93a95a16c3b2d9b3"></script>
```
Comment (ITsvetkoFF) [inline TCMSApp/src/client/index.html]: deleted by mistake?

**Comment by ITsvetkoFF** [comment]
filter is a really great feature. have no comments and corrections. your code, your unit tests. entropy has not increased. And that's good.
except few annoying (from the user's point) things.


consider the following scenario:
1. Enter `au`... you'll see the Author typeahead
2. Choose author from typeahead ... you will not see the hints: `auithor:` and `author=` (you will see hints if you enter text from the keyboard or after the next action)
so bad things are: no hints, and not too intuitive difference between `:` and `=` (we shall think about the small question sign near the filter)

And one more:
- environment
- env
- envFull

tooo many options! we shall respect our user :)

**Comment by vp9920** [comment]
Ok, we can exclude environment, but env and envFull is not identical.
About closing hints when user accept one of hints ... I know about that and I don't like it too, but I thought that this behavior impossible to change .. and yet, maybe possible .. I'll try to do something.

**Comment by vp9920** [comment]
About deleting trello and qjuery, I never deleted them, this is result of merge

**Comment by vp9920** [comment]
Not closed <section> - from master

**Comment by vp9920** [comment]
About closing hints, I tried all I could, I don't know how to change this behavior

**Comment by vp9920** [comment]
And directive template was replaced into separate file

**Comment by vp9920** [comment]
Wait a while, I wanna replace two-way data binding with one-way

**Comment by vp9920** [comment]
done

**Comment by ITsvetkoFF** [comment]
> About deleting trello and qjuery, I never deleted them, this is result of commit after resetting to updated master

but nevertheless your commit deletes that 2 lines.
and your branch has conflicts (I can't merge it) - We can check tomorrow if you wish

File: TCMSApp/src/client/app/core/constants.js [hunk 3]
```diff
@@ -6,5 +6,40 @@
         .module('app.core')
         .constant('moment', moment)
         .constant('faker', faker)
-        .constant('Trello', Trello)
+        .constant('filterFields', filterFields())
+
+    function filterFields(){
```
Comment (ITsvetkoFF) [inline TCMSApp/src/client/app/core/constants.js]: if we will have some API  constants here too this file will be too crowded. your filter settings deserve the separate file.

File: TCMSApp/karma.conf.js [hunk 4]
```diff
@@ -54,6 +54,10 @@ module.exports = function(config) {
 
         // Continuous Integration mode
         // if true, Karma captures browsers, runs the tests and exits
-        singleRun: false
+        singleRun: false,
+
+        ngHtml2JsPreprocessor: {
+            moduleName: 'templates'
+        },
```
Comment (ITsvetkoFF) [inline TCMSApp/karma.conf.js]: I see that now you have no literal with the template to inject into the test. 
Nice!
Can you please describe the solution in few details? :)

**Comment by vp9920** [comment]
filter constants at last deserved separate file

**Pull request merged**